File: Rust/209. 长度最小的子数组.rs
// 使用双指针即可， 需要注意， rust 中的类型系统十分严格， 
// usize 和 i32 之间需要有类型转换， 否则会报错
impl Solution {
    pub fn min_sub_array_len(target: i32, nums: Vec<i32>) -> i32 {
        let mut low:usize = 0;
        let mut high:usize = 0;
        let mut res:i32 = 0;
        let mut curr_sum:i32 = 0;
        while high <= nums.len(){
            if curr_sum < target {
                if high != nums.len(){
                    curr_sum = curr_sum + nums[high];
                }
                high = high + 1;
            } else{
                if res == 0 || (res!= 0 && res> (high-low) as i32){
                    res = (high - low) as i32;
                }
                curr_sum = curr_sum - nums[low];
                low = low + 1;
            }
        }
        res
    }
}

File: CPlusPlus/209. 长度最小的子数组.cpp
// 解法 1： 前缀和 + 二分 ， 需要特别注意， 后面的部分 l 超过了 nums.size() 后可能出现搜索到了边界依旧无法满足条件的情况。
#include <vector>

using namespace std;
class BinarySearch_Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums)
    {
        vector<long long> prefix { 0 };
        for (const auto& num : nums) {
            prefix.push_back(prefix.back() + num);
        }
        if (prefix.back() < target)
            return 0;
        int res = nums.size();
        for (int i = 0; i < nums.size(); ++i) {
            int base = prefix[i];
            int l = i + 1;
            int r = nums.size();
            while (l < r) {
                int mid = (l + r) >> 1;
                if (prefix[mid] - base >= target)
                    r = mid;
                else
                    l = mid + 1;
            }
            if (prefix[l] - base >= target)
                res = min(l - i, res);
        }
        return res;
    }
};
